### DISCUSSION TITLE
Failing many to one who are you

### DISCUSSION THREAD

**Discussion by orangepineapple** [ANLAB-KAIST/KENSv3#133] [Project Q&A 1. Application Layer (Socket)] [answered]
![image](https://github.com/ANLAB-KAIST/KENSv3/assets/57548877/51a8ea61-3244-4218-ac7c-bd28205c96cd)

any insight into what the many-to-ones and many to many's are testing? is it multiple clients and multiple servers? Not sure why I am getting 0.0.0.0 but I am passing the other test cases.

**Comment by Pusnow** [comment]
In many-to-one test cases, there are multiple servers and single client, and in many-to-many test cases, there are multiple servers and multiple clients.

You might use `getsockname`, `getpeername`, or `write` incorrectly.
Please check you are using right file descriptors.

Also, you can use a debugger to get insight from your code by pinpointing where the server sets the IP address to 0.0.0.0.

**Reply by orangepineapple** [reply]
Is the server ip supposed to be set to the bind_ip? because I have not found a use for the bind_ip in my code

**Reply by Pusnow** [reply]
Yes, you should use the `bind_ip` parameter for `bind()` call. The `bind()` function accepts `addr` to determine which network interface to use.

**Reply by orangepineapple** [reply]
Is the bind IP supposed to always be 0.0.0.0 
![image](https://github.com/ANLAB-KAIST/KENSv3/assets/57548877/224d0d6b-e1c2-4351-8e8b-c03218d4d44e)

I read online that this can sometimes cause the server IP to read 0.0.0.0 rather than the actual IP?
https://stackoverflow.com/questions/5759031/getsockname-always-returning-0-0-0-0

**Reply by orangepineapple** [reply]
Before this I was using INADDR_ANY which has the same behavior.

**Reply by Pusnow** [reply]
You probably use the `getsockname()` with listening socket (created with a `socket()` call).
You can retrieve the server's IP address if you call `getsockname()` with accepted socket (return value of `accept()` call).

**Reply by orangepineapple** [reply]
I am using it on the socketfd created by the accept call, if I call it on the original fd it will fail all the whoareu's but it is only failing the many to one, and many to many.

**Reply by Pusnow** [reply]
Okay, send me your code. cs341.pj@kaist.ac.kr

**Reply by orangepineapple** [reply]
Sent. Thanks

**Comment by Pusnow** [comment, marked as answer]
I found that you used `getsockname()` incorrectly.

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/Documentation/man/getsockname.2.txt#L13-L15

You must initialize the content of third parameter (`addrlen`) to size of `struct sockaddr_in`.

**Reply by orangepineapple** [reply]
OH - bruh I'm blind. Nice catch! Thanks 👍
